Implement designs for beatmap carousel v2

frenzibyte opened the pull request:
- Supersedes / closes https://github.com/ppy/osu/pull/22197
- [x] Depends on #31799 

The implementation is rewritten from the original PR to be flat, organised, and simple. Things to note:
 - Multiple components were refactored to support changing the target beatmap set for which those components function on (e.g. `DifficultySpectrumDisplay`, `TopLocalRankV2`, `UpdateBeatmapSetButtonV2`)
 - Components which were primarily used in previous song select design are copy-pasted into a separate `V2`-suffixed file as the intention (my intention) is to nuke the entire `Select` folder, so it doesn't make sense to reference or re-use components from that namespace. Also, `UpdateBeatmapSetButtonV2` is seeing a redesign very soon anyway.
 - `StarsGroupDefinition` is added temporarily to be able to implement a group panel specific for stars grouping, whether it will remain or not is unknown.
 - There may be opportunities to share code between all four/five panel classes here, but I'm not comfortable enough to tell how it should happen in a way that doesn't involve abstraction and mentally-unparsable code, so I'll leave that up to discussion.

Preview:

![CleanShot 2025-02-05 at 08 24 49](https://github.com/user-attachments/assets/92d52788-bd19-4dbb-a84e-7d1b6b2f34d1)
![CleanShot 2025-02-05 at 08 24 52](https://github.com/user-attachments/assets/9805b2db-0a78-4d46-ad35-efb6b50121b8)
![CleanShot 2025-02-05 at 08 24 24](https://github.com/user-attachments/assets/213f9589-99a0-4bef-9398-3682c1fe01ce)
![CleanShot 2025-02-06 at 02 18 11](https://github.com/user-attachments/assets/fafa63f7-e8ea-4172-80ca-be0359524b32)

frenzibyte commented on osu.Game/Screens/SelectV2/Carousel.cs:
```diff
@@ -463,27 +471,27 @@ private void refreshAfterSelection()
                 return;
             }
 
-            float spacing = SpacingBetweenPanels;
             int count = carouselItems.Count;
 
             Selection prevKeyboard = currentKeyboardSelection;
 
             // We are performing two important operations here:
-            // - Update all Y positions. After a selection occurs, panels may have changed visibility state and therefore Y positions.
             // - Link selected models to CarouselItems. If a selection changed, this is where we find the relevant CarouselItems for further use.
+            // - Update all Y positions. After a selection occurs, panels may have changed visibility state and therefore Y positions.
+
```
This change of order is necessary as the Y position now has a dependency on which beatmap is selected (see `BeatmapCarousel.GetSpacingBetweenItems`). Without that dependency, there's no way to make the expanded/selected beatmap set not overlap with the beatmap set above it, which looks broken to my eyes.

With overlap:
![CleanShot 2025-02-03 at 02 32 52](https://github.com/user-attachments/assets/9544c83e-bc49-4c28-ba3b-eff591881cc7)

Without overlap:
![CleanShot 2025-02-03 at 02 33 06](https://github.com/user-attachments/assets/9d28a489-aa8a-4c60-95c8-03ef594cd2ec)

peppy replied:
This undoes an important optimisation (you added back a second traversal of every item, which is costly), so will require scrutiny.

peppy replied:
For now please undo this one. Let's keep this PR as implementing the design and making minimal-to-no changes to the carousel code. You're welcome to force push away any changes if it will help with conflict avoidance with other carousel PRs.

peppy commented on osu.Game.Tests/Visual/SongSelectV2/TestSceneSongSelect.cs:
```diff
@@ -64,6 +104,12 @@ public override void SetUpSteps()
             AddStep("load screen", () => Stack.Push(new Screens.SelectV2.SongSelectV2()));
             AddUntilStep("wait for load", () => Stack.CurrentScreen is Screens.SelectV2.SongSelectV2 songSelect && songSelect.IsLoaded);
+
+            AddStep("download bundled beatmaps", () =>
```
Surely this wasn't supposed to stay around. Unless you're going to mark this as a non-headless test or something..

frenzibyte replied:
The intention was to make this non-headless, but I've removed it as per discussion.

peppy commented:
Going to provide an initial feedback pass on this. Let's ignore transitions / animations for now since I will do a pass on that after everything is implemented and stable.

![osu Game Tests 2025-02-04 at 06 25 04](https://github.com/user-attachments/assets/568af241-bd41-4c81-b399-05ebcb2eb13e)

The set panel is showing difficulty information but has a difficulty underneath it. I guess a post step is to remove this and only show the set panel? This doesn't really mesh with how I set things up – I was hoping the `BeatmapCarousel` would be the one doing the conditional switching, not the set header.

For now shall we ignore these details and focus on the designs? If so, I think having a test scene which shows each of the panel designs in isolation could be handy to have if it's not more than 10 minutes of work.

https://github.com/user-attachments/assets/f8816ee5-3398-4bfc-88d3-e9b22a77b953

Why is the shade of the set header here different between these two beatmaps? One is bright white, the other is gray. I guess it's the expanded flag going wrong for one reason or another?

https://github.com/user-attachments/assets/9fc13a01-b2c0-4893-83f5-909399d95b5e

Test scene doesn't seem to work

---

I don't know where to cut off review feedback with this PR. Selection state feels bad and unintuitive:

https://github.com/user-attachments/assets/8ec61af1-33a8-41b8-bd58-c9fea66b4e8a

- The shadow/glow applied to cards doesn't look great and I'm not immediately sure how to fix that (but one thing to note is you can't change the glow/shadow radius based on the expanded state without a transition it looks weird, also you can't fade a colour from black to non-black while also changing the opacity it doesn't work well).
- Is the intended direction here copy pasting the same code across three to four classes? I think this is going to be a bit of a PITA for iterating design changes.
- Is there a group panel design?

peppy commented on osu.Game/Screens/SelectV2/BeatmapPanel.cs:
```diff
@@ -79,9 +271,127 @@ protected override void PrepareForUse()
+            authorText.Text = BeatmapsetsStrings.ShowDetailsMappedBy(beatmap.Metadata.Author.Username);
 
-            this.FadeInFromZero(500, Easing.OutQuint);
+            starDifficultyBindable = null;
+            keyCount.Value = null;
+
+            updateVisibilityState();
+            isCurrentItemVisible.TriggerChange();
```
naaaaaaaaaaaaaaaaaaaaaaaaaaaah

Just do the SR calculation immediately, this is all jank as hell (no bindable, no `TriggerChange`). We can optimise the "is visible" or whatever you're trying to do here at a later point.

Please dial this PR back to be just about the design.

frenzibyte replied:
As discussed, there should be basic time-based debouncing instead of this kind of logic.

peppy replied:
I think it's fine to add that as an afterthought. It's going to be added complexity and I'd rather this PR just stays on getting the design right first.

peppy commented on osu.Game/Screens/SelectV2/BeatmapPanel.cs:
```diff
@@ -79,9 +271,127 @@ protected override void PrepareForUse()
+
+        private void updatePanelPosition()
+        {
+            if (Selected.Value)
+                panel.MoveToX(25f, duration, Easing.OutQuint);
```
This means that you can't track the keyboard selection when it's the current selection. I'd prefer that keyboard selection adds extra -x_offset in addition to selected value:


https://github.com/user-attachments/assets/39dfdc64-1324-4f55-8b42-89b9d658c553

```diff
diff --git a/osu.Game/Screens/SelectV2/BeatmapPanel.cs b/osu.Game/Screens/SelectV2/BeatmapPanel.cs
index 74f4cff3c0..1352801ea7 100644
--- a/osu.Game/Screens/SelectV2/BeatmapPanel.cs
+++ b/osu.Game/Screens/SelectV2/BeatmapPanel.cs
@@ -308,12 +308,14 @@ private void updateKeyboardSelectedDisplay()
 
         private void updatePanelPosition()
         {
+            float x = 75;
+
             if (Selected.Value)
-                panel.MoveToX(25f, duration, Easing.OutQuint);
-            else if (KeyboardSelected.Value)
-                panel.MoveToX(40f, duration, Easing.OutQuint);
-            else
-                panel.MoveToX(75f, duration, Easing.OutQuint);
+                x -= 50;
+            if (KeyboardSelected.Value)
+                x -= 35;
+
+            panel.MoveToX(x, duration, Easing.OutQuint);
         }
 
         private void updateHover()

```

peppy commented on osu.Game/Screens/SelectV2/BeatmapPanel.cs:
```diff
@@ -104,7 +425,10 @@ protected override bool OnClick(ClickEvent e)
-        public void Activated() => activationFlash.FadeOutFromOne(500, Easing.OutQuint);
+        public void Activated()
+        {
+            // activationFlash.FadeOutFromOne(500, Easing.OutQuint);
```
just leave the flash for now if we have nothing better? right now it does nothing.

peppy commented on osu.Game/Screens/SelectV2/BeatmapSetPanel.cs:
```diff
@@ -3,63 +3,283 @@
+        private CancellationTokenSource? singleDiffStarsToken;
+        private IBindable<StarDifficulty?>? singleDiffStarsBindable;
+
+        private Container panel = null!;
```
Kinda apprehensive at the number of moving pieces in this one class (number of fields and number of sets and update methods). Not sure it's an issue. Maybe better if this gets split out into a case which doesn't also attempt to handle single difficulties. Not sure yet.

frenzibyte replied:
I was considering doing that but was slightly hesitant about the amount of codelines that will be copied to have the same layout, but it was only yesterday that I realise having a separate component was your plan all along. I'll split it out.

peppy commented on osu.Game/Screens/SelectV2/Carousel.cs:
```diff
@@ -277,17 +277,27 @@ await Task.Run(async () =>
+            }
         }
 
-        private static void updateItemYPosition(CarouselItem item, ref float offset, float spacing)
```
I don't understand why this is not static.

frenzibyte replied:
Simply because of the usage of `GetSpacingBetweenItems` inside the method, there's not an easy way to pass it along unless I pass it as a delegate, which sounds overkill.

peppy replied:
I can see usage in the method above, but not this one. Maybe it was removed in a refactor?

frenzibyte replied:
Ah, that's right, but also while addressing https://github.com/ppy/osu/pull/31774#discussion_r1940576708 I've had to move the usage of that spacing method in here as well, so the static has to continue being removed.

peppy commented on osu.Game/Screens/SelectV2/BeatmapSetPanel.cs:
```diff
@@ -69,11 +289,159 @@ protected override void PrepareForUse()
+                singleDiffIcon.Show();
 
-            text.Text = $"{beatmapSetInfo.Metadata}";
+                singleDiffRank.Beatmap = singleBeatmap;
```
Set to a beatmap but never cleaned up. This means that if a pooled panel goes unused, or worse, it comes back as a non-single-difficulty, it will keep a realm subscription alive indefiniteliy.

peppy commented on osu.Game/Screens/SelectV2/BeatmapSetPanel.cs:
```diff
@@ -69,11 +289,159 @@ protected override void PrepareForUse()
+
+            titleText.Text = new RomanisableString(beatmapSet.Metadata.TitleUnicode, beatmapSet.Metadata.Title);
+            artistText.Text = new RomanisableString(beatmapSet.Metadata.ArtistUnicode, beatmapSet.Metadata.Artist);
+            updateButton.BeatmapSet = beatmapSet;
```
Likewise, this is never cleaned up in `FreeAfterUse` which means the internal tracking is not cleaned up. Probably want to clean these up.

frenzibyte replied:
I will change this to be cleaned up on `FreeAfterUse` alongside any other references, but similar to https://github.com/ppy/osu/pull/31774#discussion_r1940953285, there is nothing in the button tracking the internal state of a beatmap as far as I can see?

peppy commented on osu.Game/Screens/SelectV2/UpdateBeatmapSetButtonV2.cs:
```diff
@@ -0,0 +1,198 @@
+                beatmapSet = value;
+
+                if (IsLoaded)
+                    beatmapChanged();
```
I'm not sure this is doing what you expect it to? It doesn't seem to be updating anything when the set changes. It will still be tracking the previous set. 😅

frenzibyte replied:
I must be missing something here, this function is a callback whenever `beatmapSet` changes, and that function updates the display state of the button. There's nothing else to update..?

peppy replied:
You might be right on this one. I thought a hook-up was required to the downloader, but seems not.

frenzibyte commented on osu.Game/Screens/SelectV2/BeatmapPanel.cs:
```diff
@@ -1,75 +1,267 @@
+        private IBindable<StarDifficulty?>? starDifficultyBindable;
+        private CancellationTokenSource? starDifficultyCancellationSource;
+
+        private readonly Bindable<int?> keyCount = new Bindable<int?>();
```
That was mostly done also because of the "calculate on visibility" optimisation, or maybe I haven't thought it well. Either way that can be simplified to a single computation call in `PrepareForUse` thanks to pooling.

frenzibyte commented on osu.Game/Screens/SelectV2/BeatmapPanel.cs:
```diff
@@ -1,75 +1,267 @@
-            Size = new Vector2(500, CarouselItem.DEFAULT_HEIGHT);
-            Masking = true;
+            RelativeSizeAxes = Axes.X;
+            Width = 1.25f;
```
If you're talking about the component becoming relatively-sized, I think that makes more sense, so the panels are resized relative to the width of the carousel, which the user would be allowed to configure in their ski.

For the width number specification, it's arbitrary yes, but leaves enough extra width off-screen so it's cut off smoothly at the edge of the screen.

I will revise on this to look less silly and completely respecting the relative width of the carousel itself.

peppy replied:
Yeah, was talking about the `Width` specification. Should at least have a comment as to how you came to that number.

frenzibyte commented on osu.Game/Screens/SelectV2/BeatmapPanel.cs:
```diff
@@ -1,75 +1,267 @@
-                    Padding = new MarginPadding(5),
-                    Anchor = Anchor.CentreLeft,
-                    Origin = Anchor.CentreLeft,
+                    new BufferedContainer
```
Yeah, otherwise the coloured box behind the panel will conflict during transitions, creating an unpleasent look.

Without buffered:

https://github.com/user-attachments/assets/a708c952-d687-4d3d-8884-7804b21ace0d

With buffered:

https://github.com/user-attachments/assets/64a15f0b-ea31-479d-9684-46d62b27e89b

frenzibyte committed:
Implement beatmap set header design

frenzibyte committed:
Implement beatmap difficulty panel design

frenzibyte committed:
Implement beatmap "standalone" panel design

frenzibyte committed:
Change default carousel item header to 50px

frenzibyte committed:
Cache overlay colour provider to carousel tests

frenzibyte committed:
Extend beatmap carousel width in tests

frenzibyte committed:
Add beatmap carousel to new song select screen

frenzibyte committed:
Allow importing real beatmaps in song select test scene

frenzibyte committed:
Implement group panel design

frenzibyte committed:
Fix design tests

peppy commented:
I think you made the panels not wide enough:

![2025-02-06 00 16 38@2x](https://github.com/user-attachments/assets/787e1ba1-aea8-4a2d-9205-d955e9184d08)

Also the keyboard selection colour change is not applied to the selected item:

https://github.com/user-attachments/assets/df6f8d51-45d8-4467-9639-02c22aea8b8a

frenzibyte commented:
> I think you made the panels not wide enough:

This part is kinda intentional, the entire carousel is shifted 20px to the right in `SongSelectV2` instead of making the panel appear as cut off:
https://github.com/ppy/osu/pull/31774/files#diff-832686445c300178c6eca7c2c8bd1cee02359d36fab0985126825fcec33ffd85R52-R53

peppy commented:
I'm pretty sure that will cut off the scrollbar?

frenzibyte committed:
Fix basic code quality issues

frenzibyte committed:
Simplify some code

frenzibyte committed:
Fix carousel tests failing due to X offsets

frenzibyte committed:
Push all beatmap panels to hide their tails

frenzibyte committed:
Make `BeatmapPanel` appear hovered on keyboard selection even if selected

Was an intentional choice but appeared weird to others instead. The feedback itself probably needs changing.

frenzibyte committed:
Abstractify beatmap panel piece and update all panel implementations

frenzibyte committed:
Fix group visual test scene

frenzibyte committed:
Add V2 suffix for easier test browsing

frenzibyte committed:
Fix broken input handling with structural changes

frenzibyte committed:
Fix helper method in carousel test scene

frenzibyte commented:
I've refactored the PR once more to appropriately share code between all panel classes (see `CarouselPanelPiece`), alongside addressing both of the points above.

frenzibyte committed:
Simplify LINQ expressions to appease CI

don't ask me

peppy committed:
Merge branch 'master' into carousel-v2-implement-designs

peppy commented:
- Wrong depth:

![osu Game Tests 2025-02-12 at 10 09 50](https://github.com/user-attachments/assets/092128c4-6f5e-4771-894d-e55fc991dd7d)

- Tests failing
- Initial x position set wrong (panels fly from left unexpectedly):

https://github.com/user-attachments/assets/e9af7a20-8063-4b57-868a-7cab10994fc8

- Selected state is shared between keyboard and mouse (no hover effect as a result):

https://github.com/user-attachments/assets/3ebeb1af-17ac-4cf8-95d7-2e92a0f44989

(This can probably be fixed in a separate PR because I can hardly load this fucking page on github and am ready to rage merge it)

I'm going to ignore about 10 other UI/UX concerns and address in follow-up PRs because this PR thread is unmanageable atm.

peppy committed:
Remove unnecessary V2 suffixes

peppy committed:
Mark `Action` as `init` only

peppy committed:
Rename classes for better discoverability / grouping

peppy commented:
I feel like `CarouselItemPiece` is forced and subclassing would be better? Also there's still plenty of shared code across all the panel classes.

Also fixing the tests probably requires subclassing or moving the `Action` handling out of that class anyway.

peppy reviewed (commented):
dealing with this PR thread is frustrating as fuck (due to github being slow to load anything).

peppy reviewed (commented):
test

peppy commented:
@frenzibyte heads up i'm going to take this one over since there's no movement from your end

peppy committed:
Merge branch 'master' into carousel-v2-implement-designs

peppy committed:
User inheritance to avoid `Piece` structural nightmare

peppy committed:
Move `PanelXOffset` to `init` property rather than ctor

Feels better to me.

peppy committed:
Fix input handling not matching latest `master` logic

peppy committed:
Tidy up remaining common code

peppy commented:
@frenzibyte I've done a clean-up pass on this to attempt to make the drawables bearable to work with. Please check my changes and see if you agree.

I'll do a second review pass later today and then look to get this in. It's not perfect (and I have serious issues with the visual design language), but nothing that would impede merging this code for future adjustment.

frenzibyte committed:
Fix `PanelBeatmapStandalone` not handling selection state

frenzibyte commented:
Changes look good. I've pushed one extra change to fix a minor issue with `PanelBeatmapStandalone` display. There's also a change that can be applied to fix panels awkwardly transforming when they first appear on the screen:

https://github.com/user-attachments/assets/abd73552-bac4-4108-94e4-3d44fea61a30

The following fixes the issue by finishing any transforms pending at `PrepareForUse` state and only doing the fade-in after. Posting as a diff in case you feel it's better suited for a follow-up PR.

```diff
diff --git a/osu.Game/Screens/SelectV2/PanelBase.cs b/osu.Game/Screens/SelectV2/PanelBase.cs
index 805cbac8eb..255e40fdb9 100644
--- a/osu.Game/Screens/SelectV2/PanelBase.cs
+++ b/osu.Game/Screens/SelectV2/PanelBase.cs
@@ -27,7 +27,7 @@ public abstract partial class PanelBase : PoolableDrawable, ICarouselPanel
         private const float keyboard_active_x_offset = 25f;
         private const float active_x_offset = 50f;
 
-        private const float duration = 500;
+        protected const float DURATION = 500;
 
         protected float PanelXOffset { get; init; }
 
@@ -165,12 +165,6 @@ protected override void LoadComplete()
             KeyboardSelected.BindValueChanged(_ => updateDisplay(), true);
         }
 
-        protected override void PrepareForUse()
-        {
-            base.PrepareForUse();
-            this.FadeInFromZero(duration, Easing.OutQuint);
-        }
-
         [Resolved]
         private BeatmapCarousel? carousel { get; set; }
 
@@ -183,15 +177,15 @@ protected override bool OnClick(ClickEvent e)
 
         private void updateDisplay()
         {
-            backgroundLayer.TransformTo(nameof(Padding), backgroundLayer.Padding with { Vertical = Expanded.Value ? 2f : 0f }, duration, Easing.OutQuint);
+            backgroundLayer.TransformTo(nameof(Padding), backgroundLayer.Padding with { Vertical = Expanded.Value ? 2f : 0f }, DURATION, Easing.OutQuint);
 
             var backgroundColour = accentColour ?? Color4.White;
             var edgeEffectColour = accentColour ?? Color4Extensions.FromHex(@"4EBFFF");
 
-            backgroundAccentGradient.FadeColour(ColourInfo.GradientHorizontal(backgroundColour.Opacity(0.25f), backgroundColour.Opacity(0f)), duration, Easing.OutQuint);
-            backgroundBorder.FadeColour(backgroundColour, duration, Easing.OutQuint);
+            backgroundAccentGradient.FadeColour(ColourInfo.GradientHorizontal(backgroundColour.Opacity(0.25f), backgroundColour.Opacity(0f)), DURATION, Easing.OutQuint);
+            backgroundBorder.FadeColour(backgroundColour, DURATION, Easing.OutQuint);
 
-            TopLevelContent.FadeEdgeEffectTo(Expanded.Value ? edgeEffectColour.Opacity(0.5f) : Color4.Black.Opacity(0.4f), duration, Easing.OutQuint);
+            TopLevelContent.FadeEdgeEffectTo(Expanded.Value ? edgeEffectColour.Opacity(0.5f) : Color4.Black.Opacity(0.4f), DURATION, Easing.OutQuint);
 
             updateXOffset();
             updateHover();
@@ -207,7 +201,7 @@ private void updateXOffset()
             if (KeyboardSelected.Value)
                 x -= keyboard_active_x_offset;
 
-            this.TransformTo(nameof(Padding), new MarginPadding { Left = x }, duration, Easing.OutQuint);
+            this.TransformTo(nameof(Padding), new MarginPadding { Left = x }, DURATION, Easing.OutQuint);
         }
 
         private void updateHover()
diff --git a/osu.Game/Screens/SelectV2/PanelBeatmap.cs b/osu.Game/Screens/SelectV2/PanelBeatmap.cs
index b27e5cae14..37c57987a8 100644
--- a/osu.Game/Screens/SelectV2/PanelBeatmap.cs
+++ b/osu.Game/Screens/SelectV2/PanelBeatmap.cs
@@ -182,6 +182,9 @@ protected override void PrepareForUse()
 
             computeStarRating();
             updateKeyCount();
+
+            FinishTransforms(true);
+            this.FadeInFromZero(DURATION, Easing.OutQuint);
         }
 
         protected override void FreeAfterUse()
diff --git a/osu.Game/Screens/SelectV2/PanelBeatmapSet.cs b/osu.Game/Screens/SelectV2/PanelBeatmapSet.cs
index 5c38fe8e04..e7b8995ae3 100644
--- a/osu.Game/Screens/SelectV2/PanelBeatmapSet.cs
+++ b/osu.Game/Screens/SelectV2/PanelBeatmapSet.cs
@@ -150,6 +150,9 @@ protected override void PrepareForUse()
             updateButton.BeatmapSet = beatmapSet;
             statusPill.Status = beatmapSet.Status;
             difficultiesDisplay.BeatmapSet = beatmapSet;
+
+            FinishTransforms(true);
+            this.FadeInFromZero(DURATION, Easing.OutQuint);
         }
 
         protected override void FreeAfterUse()
diff --git a/osu.Game/Screens/SelectV2/PanelBeatmapStandalone.cs b/osu.Game/Screens/SelectV2/PanelBeatmapStandalone.cs
index 948311a86e..983ed1d550 100644
--- a/osu.Game/Screens/SelectV2/PanelBeatmapStandalone.cs
+++ b/osu.Game/Screens/SelectV2/PanelBeatmapStandalone.cs
@@ -220,6 +220,9 @@ protected override void PrepareForUse()
             difficultyLine.Show();
 
             computeStarRating();
+
+            FinishTransforms(true);
+            this.FadeInFromZero(DURATION, Easing.OutQuint);
         }
 
         protected override void FreeAfterUse()
diff --git a/osu.Game/Screens/SelectV2/PanelGroup.cs b/osu.Game/Screens/SelectV2/PanelGroup.cs
index ecb64f4797..d5f48774dc 100644
--- a/osu.Game/Screens/SelectV2/PanelGroup.cs
+++ b/osu.Game/Screens/SelectV2/PanelGroup.cs
@@ -103,6 +103,9 @@ protected override void PrepareForUse()
             GroupDefinition group = (GroupDefinition)Item.Model;
 
             titleText.Text = group.Title;
+
+            FinishTransforms(true);
+            this.FadeInFromZero(DURATION, Easing.OutQuint);
         }
     }
 }
diff --git a/osu.Game/Screens/SelectV2/PanelGroupStarDifficulty.cs b/osu.Game/Screens/SelectV2/PanelGroupStarDifficulty.cs
index 0dc5a2f365..1a9c67bd06 100644
--- a/osu.Game/Screens/SelectV2/PanelGroupStarDifficulty.cs
+++ b/osu.Game/Screens/SelectV2/PanelGroupStarDifficulty.cs
@@ -129,6 +129,9 @@ protected override void PrepareForUse()
 
             chevronIcon.Colour = contentColour;
             starCounter.Colour = contentColour;
+
+            FinishTransforms(true);
+            this.FadeInFromZero(DURATION, Easing.OutQuint);
         }
 
         private void onExpanded()
```

peppy commented:
> The following fixes the issue by finishing any transforms pending at PrepareForUse state and only doing the fade-in after. Posting as a diff in case you feel it's better suited for a follow-up PR

I intentionally removed this because it was not 100% of the way there and i don't want animations to never play (aka `FinishTransforms`. Let's leave it out for now.
